src/interface/file.rs

```diff
@@ -122,168 +127,186 @@ impl EmulatedFile {
         }?;
 
         let absolute_filename = fs::canonicalize(&path)?;
+        let rawfd = f.as_raw_fd() as i32;
 
         OPEN_FILES.insert(filename.clone());
-        let filesize = f.metadata()?.len();
+        let filesize = f.metadata()?.len() as usize; 
 
-        Ok(EmulatedFile {filename: filename, abs_filename: absolute_filename, fobj: Some(Arc::new(Mutex::new(f))), filesize: filesize as usize})
+        let mapsize = ((filesize / MAP_1GB) + 1) * MAP_1GB;
+
+        f.set_len(mapsize as u64)?;
+
+        let emfile: Vec<u8>;
+        unsafe {
+            let filemap_addr = mmap(0 as *mut c_void, mapsize, PROT_READ | PROT_WRITE, MAP_SHARED, rawfd, 0 as i64);
+            let _syncret = msync(filemap_addr, mapsize, MS_ASYNC);
+            emfile =  Vec::<u8>::from_raw_parts(filemap_addr as *mut u8, mapsize, mapsize);
+        }
 
+        Ok(EmulatedFile {filename: filename, abs_filename: absolute_filename, fobj: Arc::new(Mutex::new(Some(emfile))), realfobj: Arc::new(Mutex::new(f)), filesize: filesize, rawfd: rawfd, mapsize: mapsize})
     }
 
     pub fn close(&self) -> std::io::Result<()> {
         OPEN_FILES.remove(&self.filename);
+        unsafe {
+            let mut fobj = self.fobj.lock();
+            let map = fobj.take().unwrap();
+
+            let (oldmap_addr, _oldlen, _cap) = map.into_raw_parts();
+            munmap(oldmap_addr as *mut c_void, self.mapsize);
+        }
+
+
+        let realfobj = self.realfobj.lock();
+        realfobj.set_len(self.filesize as u64)?;
+
         Ok(())
     }
 
+    fn remap_file(&mut self) {
+        let emfile: Vec<u8>;
+
+        self.mapsize = ((self.filesize / MAP_1GB) + 1) * MAP_1GB;
+
+        let realfobj = self.realfobj.lock();
+        let _lenres = realfobj.set_len(self.mapsize as u64);
+
+        unsafe {
+            let mut fobj = self.fobj.lock();
+            let map = fobj.take().unwrap();
+
+            let (oldmap_addr, oldlen, _cap) = map.into_raw_parts();
+            let newmap_addr = mremap(oldmap_addr as *mut c_void, oldlen, self.mapsize, MREMAP_MAYMOVE);
+            let _syncret = msync(newmap_addr, self.mapsize, MS_ASYNC);
+
+            emfile =  Vec::<u8>::from_raw_parts(newmap_addr as *mut u8, self.mapsize, self.mapsize);
+        }
+
+        self.fobj = Arc::new(Mutex::new(Some(emfile)));
+    }
+
     pub fn shrink(&mut self, length: usize) -> std::io::Result<()> {
 
         if length > self.filesize { 
             panic!("Something is wrong. {} is already smaller than length.", self.filename);
         }
-        match &self.fobj {
-            None => panic!("{} is already closed.", self.filename),
-            Some(f) => { 
-                let fobj = f.lock();
-                fobj.set_len(length as u64)?;
-                self.filesize = length;         
-                Ok(())
-            }
-        }
+        let realfobj = self.realfobj.lock();
+        // realfobj.set_len(length as u64)?;
```

Comment:
Why is this still commented out, don't you want to actually truncate the backing file?